master/static/srv/enrich_task_logs.py

```diff
@@ -0,0 +1,185 @@
+import argparse
+import datetime
+import json
+import os
+import queue
+import re
+import socket
+import sys
+import threading
+import time
+from typing import Any, Dict, Iterator
+
+import backoff
+from determined.common import api
+from determined.common.api import errors
+
+# Example log message given below.
+# 2022-05-12 16:32:48,757:gc_checkpoints: [rank=0] INFO: Determined checkpoint GC, version 0.17.16-dev0
+# Below regex is used to extract the rank field from the log message.
+# Excluding empty spaces this regex matches rank in the above example as [rank=0]
+rank = re.compile(
+    "(?P<space1> ?)\[rank=(?P<rank_id>([0-9]+))\](?P<space2> ?)(?P<log>.*)"
+)
+# Below regex is used to extract the message severity from the log message.
+# Excluding empty spaces and delimiter(:) this regex matches message severity level in the above example as INFO
+level = re.compile(
+    "(?P<space1> ?)(?P<level>(DEBUG|INFO|WARNING|ERROR|CRITICAL)):(?P<space2> ?)(?P<log>.*)"
+)
+
+
+# Interval at which to force a flush.
+SHIPPER_FLUSH_INTERVAL = 1  # How often to make API calls
+
+# Full jitter time on encountering an API exception.
+SHIPPER_FAILURE_BACKOFF_SECONDS = 1
+
+# Max size of the log buffer before forcing a flush.
+LOG_BATCH_MAX_SIZE = 1000
+
+# Max size of the shipping queue before we start to apply backpressure by blocking sends. We would
+# only hit this if we got underwater by three full batches while trying to ship a batch.
+SHIP_QUEUE_MAX_SIZE = 3 * LOG_BATCH_MAX_SIZE
+
+
+class ShutdownMessage:
+    pass
+
+
+class LogCollector(threading.Thread):
+    def __init__(
+        self,
+        ship_queue: queue.Queue,
+        task_logging_metadata: Dict[str, Any],
+    ):
+        self.ship_queue = ship_queue
+        self.task_logging_metadata = task_logging_metadata
+        super().__init__()
+
+    def run(self) -> None:
+        try:
+            for line in sys.stdin:
+                print(line, flush=True)
```

Comment:
it occurs to me, the slurm logs will have the extra newlines.  Can you put `end=""` in here?

And, can you maybe leave a comment explaining that we are intentionally preserving stdin on stderr so that slurm collects the logs, in addition to shipping them to determined?

Reply:
Correct.   Will fix.